pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java

```diff
@@ -152,7 +167,12 @@ public PulsarClientImpl(ClientConfigurationData conf, EventLoopGroup eventLoopGr
         } else {
             lookup = new BinaryProtoLookupService(this, conf.getServiceUrl(), conf.getListenerName(), conf.isUseTls(), externalExecutorProvider.getExecutor());
         }
-        timer = new HashedWheelTimer(getThreadFactory("pulsar-timer"), 1, TimeUnit.MILLISECONDS);
+        if (timer == null) {
+            this.timer = new HashedWheelTimer(getThreadFactory("pulsar-timer"), 1, TimeUnit.MILLISECONDS);
+            needStopTimer = true;
+        } else {
+            this.timer = timer;
+        }
```

Comment:
@linlinnn Could you please add the test back? I noticed you have removed the unit test, it's better to cover it here to make sure the timer can be closed if the timer is not a external timer.

Reply:
sure